// react-router-loading/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findMatchingElement = exports.isLoadable = exports.isSearchDifferent = exports.isPathsEqual = exports.isPathsDifferent = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactRouter = require("react-router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var isPathsDifferent = function isPathsDifferent(first, second) {
  var _first$location, _second$location;

  return ((_first$location = first.location) === null || _first$location === void 0 ? void 0 : _first$location.pathname) !== ((_second$location = second.location) === null || _second$location === void 0 ? void 0 : _second$location.pathname);
};

exports.isPathsDifferent = isPathsDifferent;

var isPathsEqual = function isPathsEqual(first, second) {
  var _first$location2, _second$location2;

  return ((_first$location2 = first.location) === null || _first$location2 === void 0 ? void 0 : _first$location2.pathname) === ((_second$location2 = second.location) === null || _second$location2 === void 0 ? void 0 : _second$location2.pathname);
};

exports.isPathsEqual = isPathsEqual;

var isSearchDifferent = function isSearchDifferent(first, second) {
  var _first$location3, _second$location3;

  return ((_first$location3 = first.location) === null || _first$location3 === void 0 ? void 0 : _first$location3.search) !== ((_second$location3 = second.location) === null || _second$location3 === void 0 ? void 0 : _second$location3.search);
};

exports.isSearchDifferent = isSearchDifferent;

var findMatchingRoute = function findMatchingRoute(context, routes) {
  var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function (matchingElement) {};
  var match; // We use React.Children.forEach instead of React.Children.toArray().find()
  // here because toArray adds keys to all child elements and we do not want
  // to trigger an unmount/remount for two <Route>s that render the same
  // component at different URLs.

  _react["default"].Children.forEach(routes, function (child) {
    if (match == null && /*#__PURE__*/_react["default"].isValidElement(child)) {
      var _context$location;

      var path = child.props.path || child.props.from;
      match = path ? (0, _reactRouter.matchPath)((_context$location = context.location) === null || _context$location === void 0 ? void 0 : _context$location.pathname, _objectSpread(_objectSpread({}, child.props), {}, {
        path: path
      })) : context === null || context === void 0 ? void 0 : context.match;
      if (match) callback(child);
    }
  });

  return match;
};

var isLoadable = function isLoadable(context, routes) {
  var isLoadable;
  var match = findMatchingRoute(context, routes, function (matchingElement) {
    isLoadable = matchingElement.props.loading;
  });
  return match ? isLoadable : false;
};

exports.isLoadable = isLoadable;

var findMatchingElement = function findMatchingElement(context, routes) {
  var element;
  var match = findMatchingRoute(context, routes, function (matchingElement) {
    element = matchingElement;
  });
  return match ? /*#__PURE__*/_react["default"].cloneElement(element, {
    location: context.location,
    computedMatch: match
  }) : null;
};

exports.findMatchingElement = findMatchingElement;
